### ISSUE TITLE
Automatically renew vault periodic tokens

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone#2079]
Right now drone supports root tokens with vault which do not expire. We should also support period tokens which need to be automatically renewed.

Renewal should be relatively simple based on docs:

```
v.client.Auth().Token().RenewSelf(3600)
```

@jmccann what do you think about attempting renewal every 5 minutes? more frequent? less frequent?

**Comment by jmccann** [comment]
This is my first foray with vault but seems like that should be OK.  Think it allows for a low ttl token that will quickly expire if drone isn't running then.  We can always adjust later if we run into something we didn't know about.

Thanks!

**Comment by gtaylor** [comment]
While you're in there, can we make this configurable?

cc @spladug

**Comment by pantocrator27** [comment]
Is the Vault plugin functionality documented anywhere?

**Comment by bradrydzewski** [comment]
@pantocrator27 yes it is documented at http://docs.drone.io/vault-secrets/

**Comment by stephansnyt** [comment]
I'd be willing to work on some of these improvements, but I assume that the drone/vault integration is in a private repo since it's an enterprise feature. Could I (and some other people in nytm org) be granted dev access?

**Comment by bradrydzewski** [comment]
@stephansnyt thanks for the feedback, and yes to providing access. I think perhaps the best option would be to publish all of the closed source enterprise code. Please allow me some time to get this all sorted out.

**Comment by bradrydzewski** [comment]
quick update --

I am planning to publish the source code after 0.8 is released, as part of the 0.9 release cycle. There are some legal implications that I need to work through to make this happen, so bear with me :)

It was also brought to my attention that IAM roles might provide a solution to some of the concerns raised above. @stephansnyt what are your thoughts? I know @gtaylor is very interested in this capability as well.

Thanks everyone for the patience here :)

**Comment by stephansnyt** [comment]
IAM roles in what context? AFAIK "IAM roles" is not a concept that is part of drone at all.

Vault has auth backends that tie into IAM roles in AWS and GCP, that's the closest I can guess of the context. If drone is running on AWS or GCP it could auth with vault server using AWS or GCP auth backends, without having to store an auth secret.

The ideal scenario is that drone authentication to vault is totally deferred to the vault side, whether vault server or vault libs that drone imports. I think this could be done with very little additional code.

**Comment by eriklincoln** [comment]
Would be willing to help contribute on this issue as well. Our platform team prefers app roles over tokens. We've integrated Vault with our SSO for better access control across a large number of users, making it easier to revoke access vs creating individual tokens we'd have to manage separately.

**Comment by gtaylor** [comment]
We do something similar to @eriklincoln, with Vault auth happening over Okta.

**Comment by stephansnyt** [comment]
I don't mean to derail the thread any further but... @eriklincoln @gtaylor could you please elaborate? I'm interested in learning what you did.

**Comment by eriklincoln** [comment]
@stephansnyt We use Okta for our Active Directory based SSO, too. People can authenticate into Vault with their same creds as they do laptop, email, etc. There's a Vault auth integration for it: https://www.vaultproject.io/docs/auth/okta.html. Our Platform as a Service team then creates Vault policies giving access to AD groups for generic backend namespaces to control access. It makes it easier to onboard/offboard new devs in terms of giving them access to secrets. We then create the app roles for specific namespaces to give apps in production access. 

My plan is to change or local dev workflow for our web server to ask the dev to log in to Vault with their AD/Okta creds and use their Vault token to access creds locally so we don't have to share any secrets with devs. I'm just getting started with the Go Vault SDK.

**Comment by bradrydzewski** [comment]
sorry to go off topic as well. @eriklincoln could you shoot me an email at brad at drone.io so I have your email address and can send you some information? thanks

**Comment by stephansnyt** [comment]
@eriklincoln we do something very similar with github auth, mapping vault policies and secret backends to github teams, and creating app roles associated with github repos. Good to hear about others doing something similar, it seems to validate what we are doing, but it was a pain to figure out with out much guidance from hashicorp.

**Comment by rochdev** [comment]
We use Drone in GCP and would like the option to use the IAM backend as well.

Is the Vault integration stable at this point? The documentation says it's experimental and potentially unstable which could be a real problem for us (we are trying out Drone in an environment that will need to be PCI compliant)

**Comment by gtaylor** [comment]
@rochdev You probably don't want to use a static root token/appID for anything super sensitive.

**Comment by bradrydzewski** [comment]
@rochdev the current implementation is stable and we are not planning any breaking changes. Planned improvements to this feature (token refresh, IAM roles, etc) are not breaking and will not have any direct end-user impact.

**Comment by jmccann** [comment]
Just curious where this is at.

Some good points are brought up in https://github.com/drone/drone/issues/2079#issuecomment-321829094 .  I'm not sure if they all need to be implemented to complete this issue or if this should be broken into smaller issues (I'm guessing the latter).

Personally for me, while I'd love to have everything implemented per that comment ... just being able to use a token other then root token is enough for me.  Configurable refresh is nice.  Being able to change the token after startup instead of having to restart Drone if token changes is nice.  But not having to use root token would be best right now.  :)

**Comment by bradrydzewski** [comment]
It is now possible to use periodic tokens with Drone:

```
VAULT_ADDR=http://vault:8200
VAULT_TOKEN=8378e129-3228-d0ec-420d-a50c90c2ed16
VAULT_TOKEN_TTL=60m
VAULT_TOKEN_REFRESH=30m
```

We have added two new fields:

* `VAULT_TOKEN_REFRESH` is a [`time.Duration`](https://golang.org/pkg/time/#Duration) string value and controls how frequently the periodic tokens are renewed.
* `VAULT_TOKEN_TTL` is a [`time.Duration`](https://golang.org/pkg/time/#Duration) string value and controls how much the token TTL is incremented when refreshed. This value must be higher than the refresh rate and should give some margin in case the server goes down for a short period.

If correctly configured you will see the following debug logs:

```
DEBU[0000] vault: connect: http://127.0.0.1:8200        
DEBU[0000] vault: token renewal enabled: renew every 30m 
```

If the token refreshes you will see the following:

```
DEBU[0180] vault: refreshing token: increment 60s       
DEBU[0180] vault: refreshing token succeeded  
```

If the token fails to refresh you will see the following:

```
DEBU[0060] vault: refreshing token: increment 60m     
ERRO[0060] vault: refreshing token failed: Error making API request.
```

Note that this change is only available in the `:latest` tag (pending successful completion of build 1438). I will snapshot a minor release version next week once I have confirmation this feature is working as expected.

**Comment by bradrydzewski** [comment]
closing. this is now available in 0.8.5. thanks everyone!
